rotkehlchen/chain/ethereum/oracles/uniswap.py

```diff
@@ -357,40 +367,47 @@ def can_query_history(
             seconds: int | None = None,
     ) -> bool:
         """Here to comply with price historian interface."""
-        return self.is_before_contract_creation(self.ethereum.get_blocknumber_by_time(timestamp))
+        self.resolve_assets_and_load_inquirer(
+            from_asset=from_asset,
+            to_asset=to_asset,
+        )
+        return self.is_before_contract_creation(self.evm_inquirer.get_blocknumber_by_time(timestamp))  # noqa: E501
 
     def query_historical_price(
             self,
             from_asset: Asset,
             to_asset: Asset,
             timestamp: Timestamp,
     ) -> Price:
-        to_asset, from_asset = self._replace_fiat_with_stablecoin([
-            to_asset.resolve_to_asset_with_oracles(),
-            from_asset.resolve_to_asset_with_oracles(),
-        ])
+        from_token, to_token = self.resolve_assets_and_load_inquirer(
+            from_asset=from_asset,
+            to_asset=to_asset,
+        )
         return self.get_price(
-            from_asset=from_asset.resolve_to_asset_with_oracles(),
-            to_asset=to_asset.resolve_to_asset_with_oracles(),
-            block_identifier=self.ethereum.get_blocknumber_by_time(timestamp),
+            from_token=from_token,
+            to_token=to_token,
+            block_identifier=self.evm_inquirer.get_blocknumber_by_time(timestamp),
             timestamp=timestamp,
         )
 
 
 class UniswapV3Oracle(UniswapOracle):
 
-    def __init__(self, ethereum_inquirer: 'EthereumInquirer'):
-        super().__init__(ethereum_inquirer=ethereum_inquirer, version=3)
-        self.uniswap_v3_pool_abi = self.ethereum.contracts.abi('UNISWAP_V3_POOL')
-        self.uniswap_v3_factory = self.ethereum.contracts.contract(string_to_evm_address('0x1F98431c8aD98523631AE4a59f267346ea31F984'))  # noqa: E501
+    def __init__(self, evm_inquirer: 'EvmNodeInquirer'):
+        super().__init__(evm_inquirer=evm_inquirer, version=3)
+        self.uniswap_v3_pool_abi = self.evm_inquirer.contracts.abi('UNISWAP_V3_POOL')
+        self.update_factory_contract()
+
+    def update_factory_contract(self) -> None:
+        self.uniswap_v3_factory = self.evm_inquirer.contracts.contract(UNISWAPV3_FACTORY_ADDRESSES[self.evm_inquirer.chain_id])  # noqa: E501
```

Comment:
no need to load it from the database each time. make it a dict with key the chain and it should work